File: video/ai-extensions/video-80-leetcode-5/leetcode5.dart

class Solution {
  String convert(String s, int numRows) {
    if (numRows == 1) return s;

    List<StringBuffer> rows = List<StringBuffer>.generate(
        numRows, (index) => StringBuffer());
    int currentRow = 0;
    bool goingDown = false;

    for (int i = 0; i < s.length; i++) {
      rows[currentRow].write(s[i]);
      if (currentRow == 0 || currentRow == numRows - 1) {
        goingDown = !goingDown;
      }
      currentRow += goingDown ? 1 : -1;
    }

    StringBuffer result = StringBuffer();
    for (StringBuffer row in rows) {
      result.write(row.toString());
    }

    return result.toString();
  }
}

File: video/ai-extensions/video-80-leetcode-5/leetcode5.kt

class Solution {
    fun convert(s: String, numRows: Int): String {
        if (numRows == 1) return s

        val rows = Array(minOf(numRows, s.length)) { StringBuilder() }
        var curRow = 0
        var goingDown = false

        for (c in s) {
            rows[curRow].append(c)
            if (curRow == 0 || curRow == numRows - 1) goingDown = !goingDown
            curRow += if (goingDown) 1 else -1
        }

        val result = StringBuilder()
        for (row in rows) result.append(row)
        return result.toString()
    }
}
